tensorflow/core/kernels/map_stage_op.cc

```diff
@@ -0,0 +1,497 @@
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/resource_mgr.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/lib/strings/strcat.h"
+#include "tensorflow/core/platform/env.h"
+#include "tensorflow/core/platform/mutex.h"
+
+namespace tensorflow {
+
+namespace {
+
+// General Template Definition
+template <bool Ordered, typename Key, typename Data>
+struct MapTraits {};
+
+// Partially specialise for ordered
+template <typename Key, typename Data>
+struct MapTraits<true, Key, Data>
+{
+  typedef Key key_type;
+  typedef Data data_type;
+  typedef std::map<Key, Data> map_type;
+};
+
+// Partially specialise for unordered
+template <typename Key, typename Data>
+struct MapTraits<false, Key, Data>
+{
+  typedef Key key_type;
+  typedef Data data_type;
+  typedef std::unordered_map<Key, Data> map_type;
+};
+
+// Wrapper around map/unordered_map
+template <bool Ordered>
+class StagingMap : public ResourceBase
+{
+public:
+  // Public typedefs
+  typedef MapTraits<Ordered, int64, std::vector<Tensor>> map_traits;
+  typedef typename map_traits::map_type map_type;
+  typedef typename map_traits::key_type key_type;
+  typedef typename map_traits::data_type Tuple;
+
+private:
+  // Private variables
+  int capacity_;
+  mutex mu_;
+  condition_variable not_empty_;
+  condition_variable full_;
+  map_type map_ GUARDED_BY(mu_);
+
+public:
+  // public methods
+  explicit StagingMap(int capacity)
+    { capacity_ = capacity; }
```

Comment:
https://github.com/tensorflow/tensorflow/pull/9686/commits/4f8b0a6a6c27c0005c3b34760606cdb2fa81cdd3 and https://github.com/tensorflow/tensorflow/pull/9686/commits/d97bce1c3d100f4f6070890202f2a427cb04616e